ibis/backends/singlestoredb/__init__.py

```diff
@@ -0,0 +1,1089 @@
+        >>> results = cursor.fetchall()
+        >>> cursor.close()
+        >>> # Using with context manager
+        >>> with con.raw_sql("SHOW TABLES") as cursor:
+        ...     tables = [row[0] for row in cursor.fetchall()]
+        """
+        with contextlib.suppress(AttributeError):
+            query = query.sql(dialect=self.dialect)
+
+        con = self.con
+        autocommit = self._get_autocommit(con)
+
+        cursor = con.cursor()
+
+        if not autocommit:
+            con.begin()
+
+        try:
+            cursor.execute(query, **kwargs)
+        except Exception:
+            if not autocommit:
+                con.rollback()
+            cursor.close()
+            raise
+        else:
+            if not autocommit:
+                con.commit()
+            return cursor
+
+    def _get_schema_using_query(self, query: str) -> sch.Schema:
+        """Get the schema of a query result."""
+        import sqlglot as sg
+        from sqlglot import expressions as sge
+
+        from ibis import util
+        from ibis.backends.singlestoredb.converter import SingleStoreDBPandasData
+        from ibis.backends.singlestoredb.datatypes import _type_from_cursor_info
+
+        # Generate a unique alias for the subquery
+        alias = util.gen_name("query_schema")
+
+        # First try to wrap the query directly
+        # This is the most reliable approach for SingleStoreDB
+        sql = f"SELECT * FROM ({query}) AS `{alias}` LIMIT 0"
+
+        try:
+            with self.begin() as cur:
+                cur.execute(sql)
+                description = cur.description
+        except Exception as e:
+            # If the direct approach fails, try to parse and reconstruct the query
+            # This handles edge cases where the query might have syntax issues
+            try:
+                # Try parsing with SingleStore dialect first
+                parsed = sg.parse_one(query, dialect=self.dialect)
+            except Exception:
+                try:
+                    # Fallback to MySQL dialect
+                    parsed = sg.parse_one(query, dialect="mysql")
+                except Exception:
+                    # If all parsing fails, re-raise the original execution error
+                    raise e from None
+
+            # Use SQLGlot to properly construct the wrapped query
+            sql = (
+                sg.select(sge.Star())
+                .from_(parsed.subquery(sg.to_identifier(alias, quoted=True)))
+                .limit(0)
+                .sql(self.dialect)
+            )
+
+            with self.begin() as cur:
+                cur.execute(sql)
+                description = cur.description
+
+        names = []
+        ibis_types = []
+
+        for col_info in description:
+            name = col_info[0]
+            names.append(name)
+
+            # Use the detailed cursor info for type conversion
+            if len(col_info) >= 6:
```

Comment:
```suggestion
            if (len_col_info := len(col_info)) >= 6:
```

Since this value gets used repeatedly over the next few lines, might as well assign it to something and reuse?